action/protocol/staking/candidate_buckets_indexer_test.go

```diff
@@ -0,0 +1,138 @@
+package staking
+
+import (
+	"context"
+	"testing"
+
+	"github.com/iotexproject/iotex-core/db"
+	"github.com/iotexproject/iotex-proto/golang/iotextypes"
+	"github.com/stretchr/testify/require"
+	"google.golang.org/protobuf/proto"
+)
+
+func TestCandidatesBucketsIndexer_StartStop(t *testing.T) {
+	require := require.New(t)
+
+	ctx := context.Background()
+	store := db.NewMemKVStore()
+	cbi, err := NewStakingCandidatesBucketsIndexer(store)
+	require.NoError(err)
+
+	require.NoError(cbi.Start(ctx))
+	require.NoError(cbi.Stop(ctx))
+}
+
+func TestCandidatesBucketsIndexer_PutGetCandidates(t *testing.T) {
+	require := require.New(t)
+
+	ctx := context.Background()
+	store := db.NewMemKVStore()
+	cbi, err := NewStakingCandidatesBucketsIndexer(store)
+	require.NoError(err)
+
+	require.NoError(cbi.Start(ctx))
+
+	candidates := &iotextypes.CandidateListV2{}
+	candidates.Candidates = append(candidates.Candidates, &iotextypes.CandidateV2{
+		Name:               "abc",
+		SelfStakeBucketIdx: 123,
+	})
+
+	require.EqualError(cbi.PutCandidates(0, nil), "proto: Marshal called with nil")
+
+	tests := []struct {
+		height     uint64
+		candidates *iotextypes.CandidateListV2
+	}{
+		{1, &iotextypes.CandidateListV2{}},
+		{2, candidates},
+	}
+
+	for _, v := range tests {
+		require.NoError(cbi.PutCandidates(v.height, v.candidates))
+	}
+
+	//incorrect parameter
+	tests2 := []struct {
+		height        uint64
+		offset, limit uint32
+	}{
+		{0, 0, 0},
+		{2, 1111111, 1},
+		{21111111, 1111111, 1333333},
+	}
+	for _, v2 := range tests2 {
+		_, _, err = cbi.GetCandidates(v2.height, v2.offset, v2.limit)
+		require.NoError(err)
```

Comment:
add some case that would fail GetCandidates()?

Reply:
see line 83-99

Reply:
that is not what's being asked, we don't need to test after DB is closed, pls revert 83-99

what i mean is test GetCandidates() possible failure, for example:
1. when height does not exist, it returns an empty list, then you should test such
2. when offset > length, it returns an empty list, then you should test such

not only require.NoError(err)